UPGRADE-3.0.md

```diff
@@ -0,0 +1,278 @@
+                    print('$Env:PROJECT_DIRECTORY="%s"' % config['project_directory'])
+                    print('$Env:PROJECT_ROOT_DOMAIN="%s"' % config['root_domain'])
+                    print('$Env:PROJECT_DOMAINS="%s"' % domains)
+                    sys.exit(0)
+
+            if not config['power_shell']:
+                config['user_id'] = int(run('id -u', hide=True).stdout)
+
+            if config['user_id'] > 256000:
+                config['user_id'] = 1000
+
+            return config
+
+
+        locals().update(__extract_runtime_configuration(locals()))
+        ```
+
+    1. You can remove the similar part from `tasks.py` because `invoke.py` will be automatically imported.
+
+        Don't forget to remove `set_local_configuration()` function and its function call.
+
+1. Edit `tasks.py`:
+
+    1. Fix `import`s:
+
+        Remove all imports (unless you are using specific import) and add the following lines:
+
+        ```py
+        from invoke import task
+        from shlex import quote
+        from colorama import Fore
+        ```
+
+    1. Replace `run_in_docker_or_locally_for_dinghy()`,  `docker_compose()` and `docker_composer_run()` functions:
+
+        With the following content:
+
+        ```py
+        def run_in_docker_or_locally_for_dinghy(c, command, no_deps=False):
+            """
+            Mac users have a lot of problems running Yarn / Webpack on the Docker stack so this func allow them to run these tools on their host
+            """
+            if c.dinghy:
+                with c.cd(c.project_directory):
+                    c.run(command)
+            else:
+                docker_compose_run(c, command, no_deps=no_deps)
+
+
+        def docker_compose_run(c, command_name, service="builder", user="app", no_deps=False, workdir=None, port_mapping=False):
+            args = [
+                'run',
+                '--rm',
+                '-u %s' % quote(user),
+            ]
+
+            if no_deps:
+                args.append('--no-deps')
+
+            if port_mapping:
+                args.append('--service-ports')
+
+            if workdir is not None:
+                args.append('-w %s' % quote(workdir))
+
+            docker_compose(c, '%s %s /bin/sh -c "exec %s"' % (
+                ' '.join(args),
+                quote(service),
+                command_name
+            ))
+
+
+        def docker_compose(c, command_name):
+            domains = '`' + '`, `'.join([c.root_domain] + c.extra_domains) + '`'
+
+            env = {
+                'PROJECT_NAME': c.project_name,
+                'PROJECT_DIRECTORY': c.project_directory,
+                'PROJECT_ROOT_DOMAIN': c.root_domain,
+                'PROJECT_DOMAINS': domains,
+            }
+
+            cmd = 'docker-compose -p %s %s %s' % (
+                c.project_name,
+                ' '.join('-f \'' + c.root_dir + '/infrastructure/docker/' + file + '\'' for file in c.docker_compose_files),
```

Comment:
Something doesn't work well here, I get this error:

> .OSError: [Errno 22] Invalid argument: ".\\'C:\\Users\\mobma\\Dev\\docker-starter/infrastructure/docker/docker-compose.builder.yml'"

The root_dir is quite full. But if I replace it with ./ I still get an escaped `'`:

> .FileNotFoundError: [Errno 2] No such file or directory: ".\\'./infrastructure/docker/docker-compose.builder.yml'"

Using directly this works:

```python
' '.join('-f infrastructure/docker/' + file + '' for file in c.docker_compose_files),
```

Reply:
Could you try to run the command with `-e` ?
```
inv -e up
```

Reply:
Here is the result:

```
docker-compose -p app -f 'C:\Users\mobma\Dev\docker-starter/infrastructure/docker/docker-compose.builder.yml' -f 'C:\Users\mobma\Dev\docker-starter/infrastructure/docker/docker-compose.yml' -f 'C:\Users\mobma\Dev\docker-starter/infrastructure/docker/docker-compose.worker.yml' build --build-arg PROJECT_NAME=app --build-arg USER_ID=1000 php-base
```

Doesn't look so bad.

Reply:
What is strange is that I can run this command directly from Powershell and it works.

Reply:
For the record: 

The running command displayed by invoke is good :+1: 

```
docker-compose -p app -f 'C:\Users\greg\dev\docker-starter/infrastructure/docker/docker-compose.yml' -f 'C:\Users\greg\dev\docker-starter/infrastructure/docker/docker-compose.worker.yml' ps
```

If we run it manually it works.

But when it goes through invoke it does not work anymore and we got the error:

> .OSError: [Errno 22] Invalid argument: ".\'C:\Users\greg\dev\docker-starter/infrastructure/docker/docker-compose.builder.yml'"


I tried to follow the code path

1. invoke: https://github.com/pyinvoke/invoke/blob/master/invoke/runners.py#L1271 
2. Python: https://docs.python.org/3/library/subprocess.html#subprocess.Popen

So there is a mismatch here. Popen expects an array, but invoke give a string.

Anyway, regular escaping seems to work, but not in the case of a PATH. I suspect a wizard that "escape" path on windows. 

We must surround PATH with quote to avoid any issues.
So I change from single quote `'` to double quote `"` and now it works on windows and linux

If someone with strong python knowledge could explain to me what is happening here, I will be very happy.

Reply:
Single and double quotes are not interchangeable in powershell or batch. Batch requires double quotes with quite complex escaping rules (e.g. https://ss64.com/nt/syntax-esc.html). I'm not too familiar with powershell, but I believe they use back-ticks for quoting arguments.